src/orange/orangeinp/detail/BuildLogicUtils.hh

```diff
@@ -0,0 +1,365 @@
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <algorithm>
+#include <type_traits>
+#include <vector>
+
+#include "corecel/Assert.hh"
+#include "corecel/cont/VariantUtils.hh"
+#include "corecel/math/Algorithms.hh"
+#include "orange/OrangeTypes.hh"
+#include "orange/orangeinp/CsgTree.hh"
+#include "orange/orangeinp/CsgTypes.hh"
+
+namespace celeritas
+{
+namespace orangeinp
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Result of building a logic representation of a node.
+ */
+struct BuildLogicResult
+{
+    using VecLogic = std::vector<logic_int>;
+    using VecSurface = std::vector<LocalSurfaceId>;
+
+    VecSurface faces;
+    VecLogic logic;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sort the faces of a volume and remap the logic expression.
+ */
+BuildLogicResult::VecSurface remap_faces(BuildLogicResult::VecLogic& lgc)
+{
+    // Construct sorted vector of faces
+    BuildLogicResult::VecSurface faces;
+    for (auto const& v : lgc)
+    {
+        if (!logic::is_operator_token(v))
+        {
+            faces.push_back(LocalSurfaceId{v});
+        }
+    }
+
+    // Sort and uniquify the vector
+    std::sort(faces.begin(), faces.end());
+    faces.erase(std::unique(faces.begin(), faces.end()), faces.end());
+
+    // Remap logic
+    for (auto& v : lgc)
+    {
+        if (!logic::is_operator_token(v))
+        {
+            auto iter
+                = find_sorted(faces.begin(), faces.end(), LocalSurfaceId{v});
+            CELER_ASSUME(iter != faces.end());
+            v = iter - faces.begin();
+        }
+    }
+    return faces;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct a logic representation of a node.
+ *
+ * The result is a pair of vectors: the sorted surface IDs comprising the faces
+ * of this volume, and the logical representation using \em face IDs, i.e. with
+ * the surfaces remapped to the index of the surface in the face vector.
+ *
+ * The function is templated on a policy class that determines the logic
+ * representation. The policy class must have an operator() that takes a
+ * NodeId.
+ *
+ * The per-node local surfaces (faces) are sorted in ascending order of ID, not
+ * of access, since they're always evaluated sequentially rather than as part
+ * of the logic evaluation itself.
+ */
+template<class BuildLogicPolicy>
+inline BuildLogicResult build_logic(BuildLogicPolicy&& policy, NodeId n)
+{
+    static_assert(std::is_invocable_v<BuildLogicPolicy, NodeId>);
+    static_assert(std::is_rvalue_reference_v<BuildLogicPolicy&&>,
+                  "Will move from policy: rvalue ref expected");
+
+    // Construct logic vector as local surface IDs
+    auto& lgc = policy.logic();
+    CELER_EXPECT(lgc.empty());
+    policy(n);
+
+    return {remap_faces(lgc), std::move(lgc)};
```

Comment:
IIRC order of evaluation is well-defined for list initialization

Reply:
Correct, it's just for function calls that they're unspecified. This is quite subtle though...

Reply:
But I guess `std::move` doesn't really do anything anyway, it's after those two guys are evaluated that `BuildLogicResult` is constructed and `lgc`'s state changes (i.e. vector(vector&&) is called).